[Tables] Fix to correct sorting in realtime tables

akhenry opened the pull request:
Minor fix to ordering in columns. Addresses issue with how empty column values were treated in ordering and makes ordering generally a little more robust.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

larkin commented on platform/features/table/src/controllers/MCTTableController.js:
```diff
@@ -284,6 +284,20 @@ define(
+         * @returns {*} The value cast to a Number, or the original value if
+         * a Number representation is not possible.
+         */
+        MCTTableController.prototype.toNumber = function (value){
```
It looks like this is exposed externally such that it can be tested, but wouldn't it be better to test the sortComparator itself?  Seeing as this function is only utilized immediately before passing the results to `sortComparator`, it would make sense to inline this function inside of the sortComparator.  Then your test cases can focus on the expected sorting behavior instead of the implementation.

larkin commented:
This is related to https://github.com/nasa/openmct/issues/769, right?

akhenry commented:
Now also merges in changes from two other branches -

### Addresses #806

* Renamed some functions to make their purpose clearer. Also refactored where functions were mixing responsibilities (eg setElemetSizes was also setting rows)
* Addressed code review issues from #796
* Addressed need to 'set' rows before 'adding' rows. Realtime tables no longer set rows, instead using the 'add' path only.
* Reorganized Historical and Real-time controllers
* Documented MctTable directive as per #803

### Addresses #798 
* Added a new off-screen sizing table which is used to determine column widths.
* On addition of a new row, contents are assessed to see if resizing is necessary. If so, resizing of the table is performed
* Simplified markup. This removed the need for some dynamic styles. These have been replaced with statically defined styles in an external stylesheet.
* Updated tests

Because of scope of changes since last author checklist - 

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

akhenry commented:
Made some last minute changes, but all done now (I promise).

larkin commented on platform/features/table/src/TableConfiguration.js:
```diff
@@ -175,10 +185,15 @@ define(
             this.getHeaders().forEach(function (columnTitle) {
                 configuration[columnTitle] =
-                    typeof defaultConfig[columnTitle] === 'undefined' ? true :
+                    typeof (defaultConfig || {})[columnTitle] === 'undefined' ? true :
```
On line 179 you're already protecting against undefined default config; this change is redundant.

There is a lot of protection for undefined return values (breach of contract); this class should trust its private methods to follow their contract and not return undefined.  Why is it even possible that defaultColumnConfiguration can return a falsy value?

larkin commented on platform/features/table/src/controllers/MCTTableController.js:
```diff
@@ -400,74 +414,50 @@ define(
+                        largestColumn,
+                        largestColumnLength;
+                    if (!row[key]){
+                        //do nothing, no value for this column;
```
Hrm; this should fail to pass JSLint as it's an empty block.  Can you invert the truth test and remove the empty block?

larkin commented:
@akhenry Two small changes: shouldn't have to protect against undefined return types: should expect your methods to fulfill their contracts, and there's an empty if block that should be removed.  Otherwise, good to merge.  Will send back to you for small fixes and you can merge when they are resolved.

# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? If comments are resolved, yes.
4. Commit messages meet standards? Y
